src/modules/test/main.c

```diff
@@ -0,0 +1,732 @@
+#define RECIPE_UNLOAD_MODULE "UnloadModule"
+#define RECIPE_MODULE "Module"
+
+#define RECIPE_TYPE "Type"
+#define RECIPE_REPORTED "Reported"
+#define RECIPE_DESIRED "Desired"
+#define RECIPE_COMPONENT "Component"
+#define RECIPE_OBJECT "Object"
+#define RECIPE_PAYLOAD "Payload"
+#define RECIPE_JSON "Json"
+#define RECIPE_STATUS "Status"
+#define RECIPE_DELAY "Delay"
+
+#define RECIPE_RUN_COMMAND "RunCommand"
+
+typedef enum STEP_TYPE
+{
+    MODULE = 0,
+    COMMAND,
+    TEST
+} STEP_TYPE;
+
+typedef enum PAYLOAD_TYPE
+{
+    DESIRED = 0,
+    REPORTED
+} PAYLOAD_TYPE;
+
+typedef struct TEST_STEP
+{
+    PAYLOAD_TYPE type;
+    char* component;
+    char* object;
+    JSON_Value* payload;
+    int status;
+} TEST_STEP;
+
+typedef enum MODULE_ACTION
+{
+    LOAD = 0,
+    UNLOAD
+} MODULE_ACTION;
+
+typedef struct MODULE_STEP
+{
+    MODULE_ACTION action;
+    char* name;
+} MODULE_STEP;
+
+typedef struct COMMAND_STEP
+{
+    int status;
+    char* arguments;
+} COMMAND_STEP;
+
+typedef struct STEP
+{
+    STEP_TYPE type;
+    int delay;
+    union
+    {
+        MODULE_STEP module;
+        COMMAND_STEP command;
+        TEST_STEP test;
+    } data;
+} STEP;
+
+static bool g_verbose = false;
+
+void FreeStep(STEP* step)
+{
+    if (step == NULL)
+    {
+        return;
+    }
+
+    switch (step->type)
+    {
+        case MODULE:
+            free(step->data.module.name);
+            break;
+        case COMMAND:
+            free(step->data.command.arguments);
+            break;
+        case TEST:
+            free(step->data.test.component);
+            free(step->data.test.object);
+            json_value_free(step->data.test.payload);
+            break;
+    }
+}
+
+int ParseTestStep(const JSON_Object* object, TEST_STEP* test)
+{
+    int status = 0;
+    const char* type = NULL;
+    JSON_Value* payload = NULL;
+    const char* json = NULL;
+
+    if (NULL == (type = json_object_get_string(object, RECIPE_TYPE)))
+    {
+        LOG_ERROR("Missing '%s' from test step", RECIPE_TYPE);
+        status = EINVAL;
+    }
+
+    if (strcmp(type, RECIPE_REPORTED) == 0)
+    {
+        test->type = REPORTED;
+    }
+    else if (strcmp(type, RECIPE_DESIRED) == 0)
+    {
+        test->type = DESIRED;
+    }
+    else
+    {
+        status = EINVAL;
+    }
+
+    if (NULL == (test->component = strdup(json_object_get_string(object, RECIPE_COMPONENT))))
+    {
+        LOG_ERROR("'%s' is required for '%s' test step", RECIPE_COMPONENT, type);
+        status = EINVAL;
+    }
+
+    if (NULL == (test->object = strdup(json_object_get_string(object, RECIPE_OBJECT))))
+    {
+        LOG_ERROR("'%s' is required for '%s' test", RECIPE_OBJECT, type);
+        status = EINVAL;
+    }
+
+    if (NULL == (payload = json_object_get_value(object, RECIPE_PAYLOAD)))
+    {
+        if (NULL != (json = json_object_get_string(object, RECIPE_JSON)))
+        {
+            payload = json_parse_string(json);
+        }
+    }
```

Comment:
And if not (on missing else cases)? There should be errors logged

Reply:
In this case it is not an error if the payload is missing, we need to treat payloads as optional for both comparison and setting "empty" desired payloads as part of the tests.